infer/environment.cc

```diff
@@ -1690,6 +1693,15 @@ Environment::processBinding(core::Context ctx, const cfg::CFG &inWhat, cfg::Bind
         ENFORCE(tp.type != nullptr, "Inferencer did not assign type: {}", bind.value.toString(ctx, inWhat));
         tp.type.sanityCheck(ctx);
 
+        if (!tp.type.isFullyDefined()) {
+            if (auto *appliedType = core::cast_type<core::AppliedType>(tp.type)) {
+                for (auto it = appliedType->targs.begin(); it != appliedType->targs.end(); ++it) {
+                    *it =
+                        core::Types::replaceSelfType(ctx, *it, ctx.owner.enclosingClass(ctx).data(ctx)->selfType(ctx));
+                }
+            }
+        }
```

Comment:
Perhaps it would be nicer to implement `replaceSelfType` for `AppliedType`, I wasn't sure of the implications. Happy to do that if you think it's a good idea.

Reply:
Hey so this gets into this behavior:

https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%0Asig%20%7B%20returns%28T%3A%3AArray%5BT.self_type%5D%29%20%7D%0Adef%20example%0A%20%20%5B%5D%0Aend

I think that for better or worse, implementing robust support for `T.self_type` in method bodies requires a bit more of a holistic change, rather than patching around it in a handful of cases like seen in this PR.

I don't really have much more constructive to say about what might be required for that, and because we have other things that we're trying to prioritize right now I don't really have the time to provide advice here.

Unfortunately I think that means we'll have to leave `T.self_type` alone for a while.

Reply:
Fair enough, thanks for the update. I'll close this for now.